RadixWallet/Clients/FactorSourcesClient/FactorSourcesClient+Live.swift

```diff
@@ -29,30 +29,54 @@ extension FactorSourcesClient: DependencyKey {
 		}
 
 		let addPrivateHDFactorSource: AddPrivateHDFactorSource = { request in
-			let factorSource = request.factorSource
-
-			switch factorSource {
-			case let .device(deviceFactorSource):
-				try secureStorageClient.saveMnemonicForFactorSource(.init(mnemonicWithPassphrase: request.mnemonicWithPasshprase, factorSource: deviceFactorSource))
-			default:
-				loggerGlobal.notice("Saving of non device private HD factor source not permitted, kind is: \(factorSource.kind)")
+			let privateHDFactorSource = request.privateHDFactorSource
+			let deviceFactorSource = privateHDFactorSource.factorSource
+			let factorSourceID = deviceFactorSource.id
+
+			do {
+				try secureStorageClient.saveMnemonicForFactorSource(privateHDFactorSource)
+			} catch {
+				if
+					secureStorageClient.containsMnemonicIdentifiedByFactorSourceID(factorSourceID),
+					request.onMnemonicExistsStrategy == .appendWithCryptoParamaters
+				{
+					loggerGlobal.notice("Failed to save mnemonic, since it already exists, so this was expected.")
+				} else {
+					loggerGlobal.error("Failed to save mnemonic, error: \(error)")
+					throw error
+				}
 			}
-			let factorSourceID = factorSource.id
 
 			/// We only need to save olympia mnemonics into Profile, the Babylon ones
 			/// already exist in profile, and this function is used only to save the
 			/// imported mnemonic into keychain (done above).
 			if request.saveIntoProfile {
-				do {
-					try await saveFactorSource(factorSource)
-				} catch {
-					loggerGlobal.critical("Failed to save factor source, error: \(error)")
-					if let idForMnemonicToDelete = try? factorSourceID.extract(as: FactorSourceID.FromHash.self) {
+				if
+					let existingInProfile = try await getFactorSources()
+					.filter({ $0.id == factorSourceID.embed() }
+					).map({ try $0.extract(as: DeviceFactorSource.self) })
+					.first
+				{
+					switch request.onMnemonicExistsStrategy {
+					case .abort:
+						throw FactorSourceAlreadyPresent()
+					case .appendWithCryptoParamaters:
+						var updated = existingInProfile
+						loggerGlobal.critical("🔮 Appending crypto parameters to DeviceFactorSource, BEFORE: \(updated.common.cryptoParameters)....")
+						updated.common.cryptoParameters.append(request.privateHDFactorSource.factorSource.common.cryptoParameters)
+						loggerGlobal.critical("🔮 Appended crypto parameters to DeviceFactorSource, AFTER: \(updated.common.cryptoParameters) ✅")
+						try await updateFactorSource(updated.embed())
```

Comment:
if existing exist and `onExistsStrategy == . appendWithCryptoParamaters ` => we append new cryptoParams to the existing FactorSource and call `updateFactorSource`